contracts/examples/erc1155/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+		}
+
+		self.set_last_valid_type_id(&type_id);
+		self.set_token_type_uri(&type_id, uri);
+
+		// self.transfer_single_event(&caller, &from, &to, &id, &amount);
+
+		// uri event
+
+		type_id
+	}
+
+	#[endpoint]
+	fn mint(&self, type_id: BigUint, amount: BigUint) -> SCResult<()> {
+		let creator = self.get_token_type_creator(&type_id);
+
+		require!(
+			self.get_caller() == creator,
+			"Only the token creator may mint more tokens"
+		);
+
+		self.increase_balance(&creator, &type_id, &amount);
+
+		if self.get_is_fungible(&type_id) == false {
+			let last_valid_id = self.get_last_valid_token_id_for_type(&type_id);
+			let id_first = last_valid_id.clone() + BigUint::from(1u32);
+			let id_last = last_valid_id + amount;
+
+			self.set_owner_for_range(&type_id, &id_first, &id_last, &creator);
+
+			self.set_last_valid_token_id_for_type(&type_id, &id_last);
+		}
+
+		// self.transfer_single_event(&caller, &from, &to, &id, &amount);
+
+		Ok(())
+	}
+
+	#[endpoint]
+	fn burn(&self, type_id: BigUint, amount: BigUint) -> SCResult<()> {
+		require!(
+			self.get_is_fungible(&type_id) == true,
+			"Only fungible tokens can be burned"
+		);
+
+		let caller = self.get_caller();
+		let balance = self.balance_of(&caller, &type_id);
+
+		require!(balance >= amount, "Not enough tokens to burn");
+
+		self.decrease_balance(&caller, &type_id, &amount);
+
+		Ok(())
+	}
+
+	// views
+
+	#[view(supportsInterface)]
+	fn supports_interface(&self, interface_id: u32) -> bool {
+		interface_id == INTERFACE_SIGNATURE_ERC165 || interface_id == INTERFACE_SIGNATURE_ERC1155
+	}
+
+	#[view(balanceOf)]
+	fn balance_of(&self, owner: &Address, type_id: &BigUint) -> BigUint {
+		self.get_balance_mapper(&owner)
+			.get(&type_id)
+			.unwrap_or_else(|| BigUint::zero())
+	}
+
+	// returns balance for each (owner, id) pair
+	#[view(balanceOfBatch)]
+	fn balance_of_batch(&self, owners: &[Address], type_ids: &[BigUint]) -> SCResult<Vec<BigUint>> {
+		require!(
+			owners.len() == type_ids.len(),
+			"Argument lengths do not match"
+		);
+
+		let mut batch_balance = Vec::new();
+		for (owner, type_id) in owners.iter().zip(type_ids.iter()) {
+			batch_balance.push(self.balance_of(&owner, &type_id));
+		}
+
+		Ok(batch_balance)
+	}
+
+	#[view(isApprovedForAll)]
+	fn is_approval_for_all(&self, owner: Address, operator: Address) -> bool {
+		self.get_is_approved(&operator, &owner)
+	}
+
+	// private
+
+	// mock
+	fn is_smart_contract_address(&self, _address: &Address) -> bool {
+		false
+	}
+
+	fn is_valid_type_id(&self, type_id: &BigUint) -> bool {
+		type_id > &0 && type_id <= &self.get_last_valid_type_id()
+	}
+
+	fn is_valid_token_id(&self, type_id: &BigUint, token_id: &BigUint) -> bool {
+		self.is_valid_type_id(type_id)
+			&& token_id > &0
+			&& token_id <= &self.get_last_valid_token_id_for_type(type_id)
+	}
+
+	fn increase_balance(&self, owner: &Address, type_id: &BigUint, amount: &BigUint) {
+		let mut balance_mapper = self.get_balance_mapper(owner);
```

Comment:
increase_balance and decrease_balance can be expressed more simply through balance_of and set_balance

Reply:
That is slightly less efficient, since it has to construct the storage mapper twice

Reply:
true, although, in my opinion, we should focus on clarity and code re-use first (unless it's performance-critical code - which in this particular case it's not)